### ISSUE TITLE
Potential deadlock when using piece "Dummy" with promotion

### ISSUE THREAD

**Issue by jvarnholt** [thomas-maeder/popeye#510]
I found a stipulation with a potential issue in Popeye 4.89.

On my Windows-PC there is no printed output after more than one day runtime. Maybe caused by an internal endless loop or similar.
Problem seems to be the bPf2 (promotion option) because with using rook or bishop there is no issue.

The stipulation is:
anfangproblem
autor Potential Bug in Popeye 4.89
forderung 14->ser-=1
steine weiss la1 ka4 dd8 due1 duf6
schwarz duc1 dud2 due3 due8 bf2 bf7 kf8 dug5 lh6 sh7
opt zugnummern
endeproblem

Please check if this deadlock is reproducible and investigate.

**Comment by JoshuaGreen** [comment]
I've done some quick testing on this.  I agree that I don't see any output for a long time, and none of AddressSanitizer, MemorySanitizer, UBSanitizer, or our `assert`ions notice anything amiss.  However, if I add `Option intelligent` then solving finishes quickly, finding no solutions.  (Is that correct?)  Thus, I'm wondering if this could just be a slow stipulation in non-intelligent mode.

(Is `14->ser-=1` equivalent to `ser-h=14`?)

**Comment by thomas-maeder** [comment]
> (Is 14->ser-=1 equivalent to ser-h=14?)

No. It's series-stalemate in 1 (White reaches Black's stalemate) with 14 introductory series moves (by Black).

**Comment by thomas-maeder** [comment]
Hmm - so maybe it is the same thing...

**Comment by thomas-maeder** [comment]
@jvarnholt 
Can you provide the intended solution?
Or even better: can you shorten the intended the solution to, say, 3 introductory moves and still demonstrate the problem? That would be more practical for testing.

**Comment by thomas-maeder** [comment]
If I use the new option `bis` (`upto`), I get a clue:

```
opt zugnummern start 1:1:1:1 bis 40:40:40:40
endeproblem
```

```
  1  (f2-f1=D    Zeit = 0.022 s)
  1:1  (f2-f1=D    Zeit = 0.022 s)
  1:1:1  (Df1-h3    Zeit = 0.022 s)
  1:1:1:1  (Df1-h3    Zeit = 0.022 s)
```

**Comment by thomas-maeder** [comment]
NB: intelligent mode should be applicable to positions with dummies. If I add `opt intelligent`, Popeye terminates after 4.5s on my notebook.

**Comment by JoshuaGreen** [comment]
Mine continues with:
```
1:1:2  (Df1-g2    Zeit = 7:54.499 m:s)
1:1:2:1  (Df1-g2    Zeit = 7:54.499 m:s)
```

**Comment by thomas-maeder** [comment]
Now that move numbers work reasonably well for this kind of problem in the develop branch, I have made some measurements:

8->ser-=1: 12.851 s
9->ser-=1: 48.376 s
10->ser-=1: 2:26.564 m:s
11->ser-=1: 6:47.285 m:s

I tested with spinach, giving each 6 sub-processes 1GB of memory for the hash table.
Until 11->ser-=1, these 1GB were not filled by the hash table. The solving time grows linearly, with a factor of 3 to 4.

This changes drastically when we raise the length to

12->ser-=1: 2:38:20 h:m:s

because Popeye now has to do housekeeping in the hash table. It's hard to predict how solving times will grow if we try 13->ser-=1 or 14->ser-=1.


Maybe it's time to consider a suggestion made some years ago by Vaclav Kotešovec, who has a lot of experience in filling up RAM with hash tables :-) 

> I highly recommend to clear the memory (as default) when it is full.

**Comment by thomas-maeder** [comment]
In the meantime, I have managed to continue the linear series:
12->ser-=1: 17:37.651 m:s
13->ser-=1: 42:59.840 m:s
14->ser-=1: 2:51:20 h:m:s

The secret of my success was increasing the memory allowed to the Popeye processes:
12->ser-=1: 2GB
13->ser-=1: 4GB
14->ser-=1: 8GB

For 13->ser-=1, the most hungry processes consumed 2.7 of these 4GB.
And for 14->ser-=1, 5.4 of the 8GB

@jvarnholt Do you have a machine where you can allow Popeye 6GB of memory? If yes: can you reproduce what I have observed?

**Comment by thomas-maeder** [comment]
I have just tried the Kotešovec approach of entirely clearing the hash table once it's full.

This gives me a testing time of 1:05:50 h:m:s for 12->ser-=1 with 1GB of memory. That's ~2.5 times better than with the housekeeping currently used in official releases!

**Comment by thomas-maeder** [comment]
My experiments have revealed that the hash table housekeeping aren't all that expensive.
But an old flag that was set when the hash table is first compressed unnecessarily cost us a lot of time.

**Comment by thomas-maeder** [comment]
I assume that we can close this.
